posix/console.c

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/trace.h>
+#include "posix_t.h"
+
+#define DEVICE_NAME "consolefs"
+#define MAGIC 0x0b292bab
+
+typedef struct _file
+{
+    struct _oe_device base;
+    uint32_t magic;
+    oe_host_fd_t host_fd;
+} file_t;
+
+static file_t* _cast_file(const oe_device_t* device)
+{
+    file_t* file = (file_t*)device;
+
+    if (file == NULL || file->magic != MAGIC)
+        return NULL;
+
+    return file;
+}
+
+static int _consolefs_dup(oe_device_t* file_, oe_device_t** new_file_out)
+{
+    int ret = -1;
+    file_t* file = _cast_file(file_);
+    oe_host_fd_t retval = -1;
+
+    if (new_file_out)
+        *new_file_out = NULL;
+
+    if (!file || !new_file_out)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Ask the host to perform this operation. */
+    {
+        oe_errno = 0;
+
+        if (oe_posix_dup_ocall(&retval, file->host_fd) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        if (retval == -1)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Allocate and initialize a new file structure. */
+    {
+        file_t* new_file;
+
+        if (!(new_file = oe_calloc(1, sizeof(file_t))))
+            OE_RAISE_ERRNO(OE_ENOMEM);
+
+        memcpy(new_file, file, sizeof(file_t));
+        new_file->host_fd = retval;
+        *new_file_out = (oe_device_t*)new_file;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static int _consolefs_ioctl(
+    oe_device_t* file,
+    unsigned long request,
+    uint64_t arg)
+{
+    int ret = -1;
+
+    OE_UNUSED(file);
+    OE_UNUSED(request);
+    OE_UNUSED(arg);
+    OE_RAISE_ERRNO(OE_ENOTSUP);
+
+done:
+    return ret;
+}
+
+static int _consolefs_fcntl(oe_device_t* file, int cmd, uint64_t arg)
+{
+    int ret = -1;
+    OE_UNUSED(file);
+    OE_UNUSED(cmd);
+    OE_UNUSED(arg);
+    OE_RAISE_ERRNO(OE_ENOTSUP);
+
+done:
+    return ret;
+}
+
+static ssize_t _consolefs_read(oe_device_t* file_, void* buf, size_t count)
+{
+    ssize_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    oe_errno = 0;
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_read_ocall(&ret, file->host_fd, buf, count) != OE_OK)
```

Comment:
Would a check for buf!=nullptr be valid here?

Reply:
Good question Radhika. Strangely, it is legal to pass a NULL buffer to **read()** on the host side. The following is legal for example:

```
read(fd, NULL, 0); /* return zero bytes read */
```

And the following produces an **EFAULT** error:

```
read(fd, NULL, 16); /* return zero bytes read */
```